cEP-0024.md

```diff
@@ -0,0 +1,252 @@
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there are a number of missing features which would save time of
+coala community.
+
+Many of the plugins in GitMate are general plugins, but coala needs more
+plugins which support coala development workflow. So, this project
+adds plugins to auto-review, auto-reject, auto-label PRs, assign, unassign
+issues, squash commits.
+
+## Changes in coala workflow
+
+### Use increasing difficulty level order to assign issues
+
+The current assign process in coala is to check team membership of a user,
+difficulty level of the issue and assign issue to user if eligible. To
+implement the current assign process in GitMate, IGitt would need to support
+teams.
+
+Currently, IGitt doesn't support teams and there are no teams in GitLab.
+Considering [GitLab groups](https://docs.gitlab.com/ee/user/group/) and
+[GitHub teams](https://help.github.com/articles/about-teams/) as teams in
+IGitt is a possible solution for IGitt implementation. But, it requires extra
+tools to sync GitHub team and GitLab group memberships, which is a manual
+process and isn't automated yet.
+
+Also, team membership information is not publicly accessible in GitHub, so any
+system which needs to implement the current assign system would require
+organization authentication tokens.
+
+Best possible solution would be using an increasing difficulty level order,
+so that a user can be assigned to an issue only if user has solved issues of
+lower difficulty level before requesting to get assigned to a higher difficulty
+level issue.
+
+#### Drawbacks
+
+If user completed the lower difficulty level issues in GitHub and want to
+work on a higher difficulty level issue in GitLab, he/she has to repeat the
+newcomer process in GitLab. To resolve this issue, GitMate needs to find a way
+to link GitHub org and GitLab org, this would be easy when they have same name,
+as in the case of `coala`. But, resolving this issue is out of scope for this
+project.
+
+If a user is worthy enough to work on the higher level difficulty issues
+directly, then that user needs to be manually assigned by the maintainers until
+that user completes at least one low difficulty issue. Even with current assign
+system, manual intervention by maintainers is needed to add user to developers
+team.
+
+## New Features
+
+### Auto-assign issue when it is mentioned in any of PR commits
+
+When a PR is opened or updated, if any of PR's commit message referenced
+any issues with keywords `Fixes` or `Closes` and there is no
+CI failure in PR, then auto assign PR author to the mentioned unassigned
+issues.
+
+If someone else is working on the mentioned issue, comment will be added
+on the PR (configurable).
+
+If PR author is not a member of the organization, then he/she can't be assigned
+to any of the issues, a label(configurable) will be added on the PR so that
```

Comment:
This doesnt account for if the author is an org member, the issue has no assignee, but the author isnt 'eligible' for the issue's difficulty level.  i.e. hasnt done the pre-requisite newcomer or low difficulty tasks before attempting the issues of a high difficulty.

I think we at least need a separate label and message for that scenario.

> label(configurable)

Anywhere you have no space before a round bracket in this document, you need to fix the text.  Almost anywhere you have a round bracket in this document, you are not using readible text and you need to fix the text.

@sils and I both told you about this a month ago.  See https://github.com/coala/cEPs/pull/142#discussion_r190880400 .  It is ridiculous that the mentors need to find these every time.  When we ask for things to be fixed, they should be fixed everywhere in the document and the bad syntax/grammar not repeated again for us to find again.

Reply:
> I think we at least need a separate label and message for that scenario.

Label isn't needed, a message will be commented regarding what the problem is, those messages aren't configurable.

Reply:
> I think we at least need a separate label and message for that scenario.

Labelling isn't needed here and comment message when will not configurable, it will be a fixed one.

Reply:
This is completely unacceptable.  Labelling is needed.  Ineligible difficulty PRs need to be tracked otherwise our newcomer process is worthless.

Reply:
referred to at https://github.com/coala/cEPs/pull/142#discussion_r202529841